File: apps/mobile/ios/TennisScoring/RankingEngine.swift
import Foundation

struct RankingInput: Equatable { let userId, displayName, divisionId, season: String; let matchesWon, matchesLost, setsWon, setsLost, gamesWon, gamesLost: Int }
struct HeadToHead { let player1Id, player2Id: String; let player1Wins, player2Wins: Int }
struct PlayerRanking { let input: RankingInput; let matchesPlayed, gameDifferential, rank: Int; let updatedAt: Int64 }
enum RankingEngine {
  static func compute(_ inputs: [RankingInput], headToHeads: [HeadToHead], now: Int64) -> [PlayerRanking] {
    var h2h: [String: Int] = [:]
    headToHeads.forEach { h2h["\($0.player1Id):\($0.player2Id)"] = $0.player1Wins; h2h["\($0.player2Id):\($0.player1Id)"] = $0.player2Wins }
    let ordered = inputs.sorted { a, b in
      let av = [a.matchesWon,a.setsWon,a.gamesWon,a.gamesWon-a.gamesLost], bv = [b.matchesWon,b.setsWon,b.gamesWon,b.gamesWon-b.gamesLost]
      if av != bv { return av.lexicographicallyPrecedes(bv, by: >) }
      let aw = h2h["\(a.userId):\(b.userId)", default: 0], bw = h2h["\(b.userId):\(a.userId)", default: 0]
      return aw == bw ? a.displayName.localizedCompare(b.displayName) == .orderedAscending : aw > bw
    }
    return ordered.enumerated().map { PlayerRanking(input: $0.element, matchesPlayed: $0.element.matchesWon+$0.element.matchesLost, gameDifferential: $0.element.gamesWon-$0.element.gamesLost, rank: $0.offset+1, updatedAt: now) }
  }
}

File: apps/mobile/android/app/src/main/java/com/companytennisleague/app/domain/RankingEngine.kt
package com.companytennisleague.app.domain

data class RankingInput(val userId: String, val displayName: String, val divisionId: String, val season: String, val matchesWon: Int, val matchesLost: Int, val setsWon: Int, val setsLost: Int, val gamesWon: Int, val gamesLost: Int)
data class HeadToHead(val player1Id: String, val player2Id: String, val player1Wins: Int, val player2Wins: Int)
data class PlayerRanking(val input: RankingInput, val matchesPlayed: Int, val gameDifferential: Int, val rank: Int, val updatedAt: Long)

object RankingEngine {
  fun compute(inputs: List<RankingInput>, headToHeads: List<HeadToHead>, now: Long = System.currentTimeMillis()): List<PlayerRanking> {
    val h2h = buildMap<Pair<String,String>,Int> { headToHeads.forEach { put(it.player1Id to it.player2Id,it.player1Wins); put(it.player2Id to it.player1Id,it.player2Wins) } }
    val sorted = inputs.sortedWith { a,b ->
      compareValuesBy(b,a,{it.matchesWon},{it.setsWon},{it.gamesWon},{it.gamesWon-it.gamesLost}).takeIf { it != 0 }
        ?: ((h2h[b.userId to a.userId] ?: 0) - (h2h[a.userId to b.userId] ?: 0)).takeIf { it != 0 }
        ?: a.displayName.compareTo(b.displayName)
    }
    return sorted.mapIndexed { index, it -> PlayerRanking(it, it.matchesWon + it.matchesLost, it.gamesWon - it.gamesLost, index + 1, now) }
  }
}
